packages/snjs/lib/services/features_service.ts

```diff
@@ -0,0 +1,178 @@
+  ) {
+    super();
+
+    this.removeApiServiceObserver = apiService.addEventObserver(
+      async (eventName, data) => {
+        if (eventName === ApiServiceEvent.MetaReceived) {
+          const { userUuid, userRoles } = data as MetaReceivedData;
+          await this.updateRoles(
+            userUuid,
+            userRoles.map((role) => role.name)
+          );
+        }
+      }
+    );
+
+    this.removeWebSocketsServiceObserver = webSocketsService.addEventObserver(
+      async (eventName, data) => {
+        if (eventName === WebSocketsServiceEvent.UserRoleMessageReceived) {
+          const {
+            payload: { userUuid, currentRoles },
+          } = data as UserRolesChangedEvent;
+          await this.setRoles(currentRoles);
+          await this.updateFeatures(userUuid);
+        }
+      }
+    );
+  }
+
+  public async loadUserRoles(): Promise<void> {
+    this.roles =
+      (await this.storageService.getValue(StorageKey.UserRoles)) || [];
+  }
+
+  public async updateRoles(
+    userUuid: UuidString,
+    roles: RoleName[]
+  ): Promise<void> {
+    const userRolesChanged = this.haveRolesChanged(roles);
+    if (userRolesChanged) {
+      await this.setRoles(roles);
+      await this.updateFeatures(userUuid);
+    }
+  }
+
+  private async setRoles(roles: RoleName[]): Promise<void> {
+    this.roles = roles;
+    await this.storageService.setValue(StorageKey.UserRoles, this.roles);
+  }
+
+  private haveRolesChanged(roles: RoleName[]): boolean {
+    return (
+      roles.some((role) => !this.roles.includes(role)) ||
+      this.roles.some((role) => !roles.includes(role))
+    );
+  }
+
+  private async updateFeatures(userUuid: UuidString): Promise<void> {
+    const featuresResponse = await this.apiService.getUserFeatures(userUuid);
+    if (!featuresResponse.error && featuresResponse.data) {
+      await this.mapFeaturesToItems(
+        (featuresResponse as UserFeaturesResponse).data.features
+      );
+    }
+  }
+
+  private createItemDataForFeature(feature: Feature) {
+    return {
+      content_type: feature.contentType,
+      content: {
+        identifier: feature.identifier,
+        name: feature.name,
+        hosted_url: feature.url,
+        url: feature.url,
+        local_url: null,
+        area: feature.area,
+        package_info: feature,
+        valid_until: feature.expiresAt,
+      },
+      references: [],
+    };
+  }
+
+  private async mapFeaturesToItems(features: Feature[]): Promise<void> {
+    const currentItems = this.itemManager.getItems([
+      ContentType.Component,
+      ContentType.Theme,
+    ]);
+    const itemsToDeleteUuids = [];
+
+    const today = new Date();
+
+    for (const feature of features) {
+      const expired = feature.expiresAt! < today.getTime();
+      const itemData = this.createItemDataForFeature(feature);      
+      const existingItem = currentItems.find((item) => {
+        if (
+          item.content &&
+          typeof item.content !== 'string' &&
```

Comment:
You could replace this whole conditional with `item.safeContent.package_info` I think.